### ISSUE TITLE
Instance volume switching to Read-only while processing

### ISSUE THREAD

**Issue by arielsvn** [AlexsLemonade/refinebio#2237]
### Context

We are trying to re-process `~30k` SRA samples for https://github.com/AlexsLemonade/refinebio/issues/2006#issuecomment-613538306, some need to download large files (`>10gb`).

Downloader jobs were failing with out-of-memory errors so we introduced https://github.com/AlexsLemonade/refinebio/pull/2223 to stream downloads. But the number of Out of memory errors increased, so we tried https://github.com/AlexsLemonade/refinebio/pull/2230 to re-try downloads within the same job if any error was raised and sleep in between re-tries.

Also, added some changes to the script that mounts the volume https://github.com/AlexsLemonade/refinebio/pull/2222 because it wasn't correctly identified. As well as https://github.com/AlexsLemonade/refinebio/pull/2227 because the volume wasn't being mounted when the instance started. That didn't solved it, and we have an issue for it https://github.com/AlexsLemonade/refinebio/issues/2234

### Problem or idea

Now we are having another issue, the instance is switching to readonly and we are getting the following error on the downloader jobs:

```
OSError: [Errno 30] Read-only file system: '/home/user/data_store/ERP006862'
```

Also on the instance:

```
ubuntu@ip-10-0-84-77:/var/ebs$ touch xx
touch: cannot touch 'xx': Read-only file system
ubuntu@ip-10-0-84-77:/var/ebs$ grep "[[:space:]]ro[[:space:],]" /proc/mounts 
tmpfs /sys/fs/cgroup tmpfs ro,nosuid,nodev,noexec,mode=755 0 0
/dev/xvdf /var/ebs ext4 ro,relatime,data=ordered 0 0
ubuntu@ip-10-0-84-77:/var/ebs$ df -h
Filesystem      Size  Used Avail Use% Mounted on
udev            481G     0  481G   0% /dev
tmpfs            97G  1.1G   95G   2% /run
/dev/xvda1       97G  2.4G   95G   3% /
tmpfs           481G     0  481G   0% /dev/shm
tmpfs           5.0M     0  5.0M   0% /run/lock
tmpfs           481G     0  481G   0% /sys/fs/cgroup
/dev/xvdf       2.0T  459G  1.4T  25% /var/ebs
tmpfs            97G     0   97G   0% /run/user/1000
```

### Solution or next step

by @cgreene 

> i'm guessing what's happening is a node is getting cycled, the disks are busy, aws ends up cycling the node anyway while the disk is still mounted and writing
> when we get a new node and attach the ebs it finds that it unmounted uncleanly and needs to be repaired
> that repair either fails or we're not running it
> i'm guessing this is happening for the same reasons you had so much trouble unmounting the EBS storage when you were debugging by hand

**Comment by arielsvn** [comment]
We ran an instance `c5d.9xlarge` over the last 24h with a 900GB NVM drive and it didn't raise that error. The success rate of downloader jobs was good.

![image](https://user-images.githubusercontent.com/1882507/79642178-0c3c0d00-816a-11ea-9845-50ed7897e891.png)

However, the number of jobs that we ran in parallel was way smaller, since this instance has less RAM. I never saw more than 10 downloader jobs running concurrently.

I'm going to switch the disk to the EBS volume and see if the error happens again.

**Comment by cgreene** [comment]
Good testing! Sounds like a plan to me

**Comment by arielsvn** [comment]
Looks like the EBS volume switched to readonly yesterday

```bash
ubuntu@ip-10-0-106-192:/var/ebs$ touch file
touch: cannot touch 'file': Read-only file system
ubuntu@ip-10-0-106-192:/var/ebs$ grep "[[:space:]]ro[[:space:],]" /proc/mounts 
tmpfs /sys/fs/cgroup tmpfs ro,nosuid,nodev,noexec,mode=755 0 0
/dev/nvme2n1 /var/ebs ext4 ro,relatime,data=ordered 0 0
/dev/nvme2n1 /var/ebs ext4 ro,relatime,data=ordered 0 0
```

![image](https://user-images.githubusercontent.com/1882507/79693411-791dd880-8238-11ea-878a-7fd6aeba28d9.png)

Also looks like the instance has been up all the time: `Launch time
April 17, 2020 at 3:16:18 PM UTC-4 (45 hours)`

I'm going to scale down, looks like the error is related to the EBS volume. We can discuss next steps tomorrow.

**Comment by arielsvn** [comment]
#2243 started using SSD disks on the instances, that disk also went read-only last night while we were processing:

```
ubuntu@ip-10-0-20-250:/var/ebs$ touch xx
touch: cannot touch 'xx': Read-only file system

ubuntu@ip-10-0-20-250:/var/ebs$ lsblk
NAME    MAJ:MIN RM  SIZE RO TYPE MOUNTPOINT
xvda    202:0    0  900G  0 disk 
└─xvda1 202:1    0  900G  0 part /
xvdb    202:16   0  1.8T  0 disk 

ubuntu@ip-10-0-20-250:/var/ebs$  grep "[[:space:]]ro[[:space:],]" /proc/mounts 
/dev/xvda1 / ext4 ro,relatime,discard,data=ordered 0 0
tmpfs /sys/fs/cgroup tmpfs ro,nosuid,nodev,noexec,mode=755 0 0

ubuntu@ip-10-0-20-250:/var/ebs$ df -h
Filesystem      Size  Used Avail Use% Mounted on
udev            481G     0  481G   0% /dev
tmpfs            97G  562M   96G   1% /run
/dev/xvda1      873G  441G  432G  51% /
tmpfs           481G     0  481G   0% /dev/shm
tmpfs           5.0M     0  5.0M   0% /run/lock
tmpfs           481G     0  481G   0% /sys/fs/cgroup
tmpfs            97G     0   97G   0% /run/user/1000
```

**Comment by wvauclain** [comment]
We just had the issue again with the upgraded AMI, but this time it happened while we were watching so we could grab the kernel logs. Here are the relevant ones:

```
Jul 09 13:51:05 ip-10-0-93-61 kernel: slab_out_of_memory: 789 callbacks suppressed
Jul 09 13:51:05 ip-10-0-93-61 kernel: SLUB: Unable to allocate memory on node -1 (gfp=0x2408040)
Jul 09 13:51:05 ip-10-0-93-61 kernel:   cache: kmalloc-192(20387:5a4608030568c1dd90a311dfc2a5845da6a502a468777817012fc2b1d9c54b5d), object size: 192, buffer s
Jul 09 13:51:05 ip-10-0-93-61 kernel:   node 0: slabs: 13, objs: 525, free: 0
Jul 09 13:51:05 ip-10-0-93-61 kernel:   node 1: slabs: 6, objs: 252, free: 0
Jul 09 13:51:05 ip-10-0-93-61 kernel: Aborting journal on device xvda1-8.
Jul 09 13:51:05 ip-10-0-93-61 kernel: EXT4-fs error (device xvda1) in ext4_do_update_inode:4711: Journal has aborted
Jul 09 13:51:05 ip-10-0-93-61 kernel: EXT4-fs error (device xvda1) in ext4_do_update_inode:4711: Journal has aborted
Jul 09 13:51:05 ip-10-0-93-61 kernel: EXT4-fs error (device xvda1) in ext4_setattr:5009: Out of memory
Jul 09 13:51:05 ip-10-0-93-61 kernel: EXT4-fs error (device xvda1): mpage_map_and_submit_extent:2343: comm python3: Failed to mark inode 22528169 dirty
Jul 09 13:51:05 ip-10-0-93-61 containerd[22626]: time="2020-07-09T13:51:05.963661328Z" level=info msg="shim reaped" id=607a0c7c038e4c75c98101bdf3bb792b10fb28b
Jul 09 13:51:05 ip-10-0-93-61 containerd[22626]: time="2020-07-09T13:51:05.973376662Z" level=error msg="failed to delete bundle" error="unlinkat /var/lib/cont
Jul 09 13:51:05 ip-10-0-93-61 cloud-init[2477]: time="2020-07-09T13:51:05.973825056Z" level=info msg="ignoring event" module=libcontainerd namespace=moby topi
Jul 09 13:51:05 ip-10-0-93-61 kernel: EXT4-fs error (device xvda1): ext4_journal_check_start:56:
Jul 09 13:51:05 ip-10-0-93-61 kernel: EXT4-fs error (device xvda1): ext4_journal_check_start:56:
Jul 09 13:51:05 ip-10-0-93-61 kernel: Detected aborted journal
Jul 09 13:51:05 ip-10-0-93-61 kernel:
Jul 09 13:51:05 ip-10-0-93-61 kernel: EXT4-fs error (device xvda1): mpage_map_and_submit_extent:2343: comm kworker/u258:1: Failed to mark inode 22528189 dirty
Jul 09 13:51:05 ip-10-0-93-61 kernel: EXT4-fs (xvda1): Remounting filesystem read-only
Jul 09 13:51:05 ip-10-0-93-61 kernel: EXT4-fs error (device xvda1) in ext4_writepages:2645: IO failure
Jul 09 13:51:05 ip-10-0-93-61 kernel: EXT4-fs error (device xvda1) in ext4_writepages:2645: IO failure
Jul 09 13:51:05 ip-10-0-93-61 kernel: EXT4-fs (xvda1): ext4_writepages: jbd2_start: 9223372036854729727 pages, ino 22528165; err -30
Jul 09 13:51:05 ip-10-0-93-61 kernel: EXT4-fs error (device xvda1) in ext4_writepages:2645: IO failure
Jul 09 13:51:05 ip-10-0-93-61 kernel: SLUB: Unable to allocate memory on node -1 (gfp=0x2080020)
Jul 09 13:51:05 ip-10-0-93-61 kernel:   cache: kmalloc-128(20364:46c753866dc6e1d3aa1d04d7fb9b60cf3c7b3e4a5299cfbce35dda26cc984b83), object size: 128, buffer s
Jul 09 13:51:05 ip-10-0-93-61 kernel:   node 0: slabs: 6, objs: 384, free: 0
Jul 09 13:51:05 ip-10-0-93-61 kernel:   node 1: slabs: 13, objs: 832, free: 0
Jul 09 13:51:05 ip-10-0-93-61 kernel: SLUB: Unable to allocate memory on node -1 (gfp=0x2080020)
Jul 09 13:51:05 ip-10-0-93-61 kernel:   cache: kmalloc-128(20364:46c753866dc6e1d3aa1d04d7fb9b60cf3c7b3e4a5299cfbce35dda26cc984b83), object size: 128, buffer s
Jul 09 13:51:05 ip-10-0-93-61 kernel:   node 0: slabs: 6, objs: 384, free: 0
Jul 09 13:51:05 ip-10-0-93-61 kernel:   node 1: slabs: 13, objs: 832, free: 0
Jul 09 13:51:05 ip-10-0-93-61 kernel: Detected aborted journal
Jul 09 13:51:05 ip-10-0-93-61 kernel:
Jul 09 13:51:05 ip-10-0-93-61 cloud-init[2477]: tee: /var/log/cloud-init-output.log: Read-only file system
```

Some issues on the various docker repos seem to point to the fact that with 3.x and 4.x kernels running docker with the `--memory` option would cause the kernel slab allocator to fill up, which appears to be what's happening here. We are going to try upgrading to 18.04 which has Linux 5.3.

**Comment by kurtwheeler** [comment]
We ran several days without going read-only!! It seems upgrading the kernel worked!
